cairo/ethereum/cancun/vm/precompiled_contracts/blake2f.cairo

```diff
@@ -0,0 +1,519 @@
+// SPDX-License-Identifier: MIT
+
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.cairo_builtins import HashBuiltin, KeccakBuiltin, BitwiseBuiltin
+from starkware.cairo.common.registers import get_fp_and_pc, get_label_location
+from starkware.cairo.common.bool import FALSE
+from ethereum.exceptions import EthereumException
+from ethereum.cancun.vm.exceptions import InvalidParameter
+from ethereum.cancun.vm.gas import charge_gas, GasConstants
+from ethereum.cancun.vm.evm_impl import Evm
+from ethereum_types.bytes import Bytes, BytesStruct
+from ethereum.cancun.vm.evm_impl import EvmImpl
+from ethereum.utils.numeric import divmod
+from ethereum_types.numeric import Uint
+
+from legacy.utils.utils import Helpers
+
+func blake2f{range_check_ptr, bitwise_ptr: BitwiseBuiltin*, keccak_ptr: KeccakBuiltin*, evm: Evm}(
+    ) -> EthereumException* {
+    alloc_locals;
+    const rounds_bytes_len = 4;
+    const word_bytes_len = 8;
+    const h_bytes_offset = 4;
+    const m_bytes_offset = 68;
+    const t_bytes_offset = 196;
+    const f_bytes_offset = 212;
+
+    let input = evm.value.message.value.data.value.data;
+    let input_len = evm.value.message.value.data.value.len;
+
+    // Check input length
+    if (input_len != 213) {
+        tempvar err = new EthereumException(InvalidParameter);
+        return err;
+    }
+
+    let rounds = Helpers.bytes_to_felt(rounds_bytes_len, input);
+    let err = charge_gas(Uint(GasConstants.GAS_BLAKE2_PER_ROUND * rounds));
+    if (err != cast(0, EthereumException*)) {
+        return err;
+    }
+
+    // Check the flag
+    tempvar f = input[f_bytes_offset];
+    if (f != f * f) {
+        tempvar err = new EthereumException(InvalidParameter);
+        return err;
+    }
+
+    let (local h: felt*) = alloc();
+    Helpers.load_64_bits_array(8, input + h_bytes_offset, h);
+
+    let (m: felt*) = alloc();
+    Helpers.load_64_bits_array(16, input + m_bytes_offset, m);
+
+    let t0 = Helpers.bytes_to_64_bits_little_felt(input + t_bytes_offset);
+    let t1 = Helpers.bytes_to_64_bits_little_felt(input + t_bytes_offset + 8);
```

Comment:
eventually we would want to use `Bytes_to_be_ListBytes4` instead - see sha256.cairo precompile for ref, im not strongly opinionated yet, as you wish